### ISSUE TITLE
isinstance(x, int) could be a trap in Python 2

### ISSUE THREAD

**Issue by gvanrossum** [python/mypy#5155]
In Python 2, at runtime, if `x` is a `long`, `isinstance(x, int)` returns False. But mypy treats `long` as a type alias for `int`, so mypy doesn't see that an `else` branch could be taken where `x` is still (what it considers) an `int`. E.g.
```py
def foo(a: Union[int, str]):
    if isinstance(x, int):
        # case for int
    else:
        # case for str
```
Mypy believes that in the `else` branch `x` is a `str`, but that's not true at runtime -- it could be a `long`.

Perhaps mypy should warn when it draws a conclusion like this.

**Comment by sid-kap** [comment]
Why does mypy treat `long` as a type alias for `int`?

**Comment by gvanrossum** [comment]
> Why does mypy treat long as a type alias for int?

Because the two types are *nearly* always interchangeable, and behave *nearly* identically. (And because in Python 3 there is only one type, `int`.)

**Comment by hauntsaninja** [comment]
What is Python 2?
